### PR TITLE
[Fix] Malformed attributes cause ClassCastException in XML Parser

### PR DESCRIPTION
**Pull request by comtef**
# Context
This PR fixes the issue related to malformed attributes in XML documents.

# Fix
## Schema
The schema inferring will not stop on a malformed attribute
An empty Schema will be returned if the parsing mode is not permissive
## Parsing
Documents with invalid attributes will be treated or dropped based on parsing mode

### CODE REVIEW

**Comment by comtef** [comment]
Could someone review this please?

**Comment by HyukjinKwon** [comment]
Sorry I was super busy and forgot about it ... Will try to take a look or make a proper fix soon.

File: src/test/java/com/databricks/spark/xml/JavaXmlSuite.java [hunk 1]
```diff
@@ -37,8 +40,10 @@
     private String tempDir = "target/test/xmlData/";
 
     @Before
-    public void setUp() {
-        sqlContext = new SQLContext(new SparkContext("local[2]", "JavaXmlSuite"));
+    public void setUp() throws IOException {
+        SparkConf conf = new SparkConf().set("spark.sql.warehouse.dir",
+                Files.createTempDirectory("spark-warehouse").toString());
```
Comment (HyukjinKwon) [inline src/test/java/com/databricks/spark/xml/JavaXmlSuite.java]: Is this change required?

File: src/test/scala/com/databricks/spark/xml/XmlSuite.scala [hunk 2]
```diff
@@ -17,18 +17,17 @@ package com.databricks.spark.xml
 
 import java.io.File
 import java.nio.charset.UnsupportedCharsetException
+import java.nio.file.Files
 import java.sql.{Date, Timestamp}
 
 import scala.io.Source
-
```
Comment (HyukjinKwon) [inline src/test/scala/com/databricks/spark/xml/XmlSuite.scala]: Let's keep the import order - https://github.com/databricks/scala-style-guide#imports

File: src/test/scala/com/databricks/spark/xml/XmlSuite.scala [hunk 3]
```diff
@@ -75,7 +75,9 @@ class XmlSuite extends FunSuite with BeforeAndAfterAll {
 
   override protected def beforeAll(): Unit = {
     super.beforeAll()
-    sqlContext = new SQLContext(new SparkContext("local[2]", "XmlSuite"))
+    val conf = new SparkConf().set("spark.sql.warehouse.dir",
+      Files.createTempDirectory("spark-warehouse").toString)
```
Comment (HyukjinKwon) [inline src/test/scala/com/databricks/spark/xml/XmlSuite.scala]: ditto

File: src/main/scala/com/databricks/spark/xml/parsers/StaxXmlParser.scala [hunk 4]
```diff
@@ -90,6 +90,9 @@ private[xml] object StaxXmlParser {
             failedRecord(xml)
           case _: XMLStreamException =>
             failedRecord(xml)
+          case _: ClassCastException =>
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/xml/parsers/StaxXmlParser.scala]: How about  just catching `RuntimeException` and deduplicate some exceptions here?

Comment (comtef) [reply src/main/scala/com/databricks/spark/xml/parsers/StaxXmlParser.scala]: Changed all exception handling to :

case NonFatal(e) =>
            failedRecord(xml)

File: src/test/java/com/databricks/spark/xml/JavaXmlSuite.java (hunk 1 above)
Comment (comtef) [reply src/test/java/com/databricks/spark/xml/JavaXmlSuite.java]: I worked on both Mac and Windows and Spark 2.0.0 is affected by this issue on Windows :

java.net.URISyntaxException: Relative path in absolute URI: file:C:/Users/XXXXX/IdeaProjects/spark-xml/spark-warehouse

see https://issues.apache.org/jira/browse/SPARK-15893

This is the known workaround for this issue. I added this configuration to be able to build the project on all platforms.

File: src/test/scala/com/databricks/spark/xml/XmlSuite.scala (hunk 2 above)
Comment (comtef) [reply src/test/scala/com/databricks/spark/xml/XmlSuite.scala]: Fixed, updated my IDE to follow guidelines

**Comment by comtef** [comment]
@HyukjinKwon Do you have time to review my changes?

File: src/main/scala/com/databricks/spark/xml/util/InferSchema.scala [hunk 5]
```diff
@@ -104,6 +104,11 @@ private[xml] object InferSchema {
             Some(StructType(Seq(StructField(options.columnNameOfCorruptRecord, StringType))))
           case _: XMLStreamException =>
             None
+          case _: ClassCastException if shouldHandleCorruptRecord =>
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/xml/util/InferSchema.scala]: Could we do the similar thing with `NonFatal`?

File: src/test/scala/com/databricks/spark/xml/XmlSuite.scala [hunk 6]
```diff
@@ -17,21 +17,21 @@ package com.databricks.spark.xml
...
+import org.scalatest.{BeforeAndAfterAll, FunSuite}
 
-import org.apache.spark.{SparkException, SparkContext}
-import org.apache.spark.sql.{SaveMode, Row, SQLContext}
-import org.apache.spark.sql.types._
 import com.databricks.spark.xml.XmlOptions._
 import com.databricks.spark.xml.util.ParseModes
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{Row, SQLContext, SaveMode}
+import org.apache.spark.{SparkConf, SparkContext, SparkException}
```
Comment (HyukjinKwon) [inline src/test/scala/com/databricks/spark/xml/XmlSuite.scala]: Let's keep the import order according to https://github.com/databricks/scala-style-guide#imports

File: src/test/java/com/databricks/spark/xml/JavaXmlSuite.java (hunk 1 above)
Comment (HyukjinKwon) [reply src/test/java/com/databricks/spark/xml/JavaXmlSuite.java]: Okay, let's leave some comments and make this double-space indentation.

File: src/test/scala/com/databricks/spark/xml/XmlSuite.scala (hunk 3 above)
Comment (HyukjinKwon) [reply src/test/scala/com/databricks/spark/xml/XmlSuite.scala]: Here too. Let's leave some comments and make this as:

```
val conf = new SparkConf().set(
  "spark.sql.warehouse.dir",
  Files.createTempDirectory("spark-warehouse").toString)
```

**Comment by HyukjinKwon** [comment]
Looks good otherwise. Thanks for fixing this @comtef. 👍

File: src/test/scala/com/databricks/spark/xml/XmlSuite.scala (hunk 6 above)
Comment (comtef) [reply src/test/scala/com/databricks/spark/xml/XmlSuite.scala]: Fixed

File: src/main/scala/com/databricks/spark/xml/util/InferSchema.scala (hunk 5 above)
Comment (comtef) [reply src/main/scala/com/databricks/spark/xml/util/InferSchema.scala]: Fixed

File: src/test/scala/com/databricks/spark/xml/XmlSuite.scala (hunk 3 above)
Comment (comtef) [reply src/test/scala/com/databricks/spark/xml/XmlSuite.scala]: Fixed

File: src/test/java/com/databricks/spark/xml/JavaXmlSuite.java (hunk 1 above)
Comment (comtef) [reply src/test/java/com/databricks/spark/xml/JavaXmlSuite.java]: Comment added.
This is a java file, it s indented with 4 spaces.

**Comment by comtef** [comment]
@HyukjinKwon Should be good now. I also squashed the commits.

**Comment by HyukjinKwon** [comment]
@comtef, I merged this into master.

BTW, I think you could add `COMTE FABIEN <f.comte-externe@probtp.com>` in your account in Github so that the commit log is properly linked to you.

**Comment by comtef** [comment]
@HyukjinKwon Oh, I forgot to check which account was used for the commit, it was supposed to be comtef@gmail.com...
